### PR TITLE
NTGR-640: Add privacy API for GDPR

### PR DESCRIPTION
**Pull request by kei178**
This PR adds the privacy API which is required for Moodle plugins to allow a Moodle installation to become GDPR compliant.

Documentation: https://docs.moodle.org/dev/Privacy_API

## Requirements

Our plugin does not store any personal data in the Moodle database but sends user's name and email to Accredible to issue credentials. According to "[My plugin only sends data to an external location, but doesn’t store it locally in Moodle - what should I do?](https://docs.moodle.org/dev/Subject_Access_Request_FAQ#My_plugin_only_sends_data_to_an_external_location.2C_but_doesn.E2.80.99t_store_it_locally_in_Moodle_-_what_should_I_do.3F)", our plugin needs to define:

1. `get_metadata` with `add_external_location_link()` to indicate that the plugin sends data to the external system (Accredible).
2. `core_privacy/local/request/plugin/provider` with empty methods including:
    - get_contexts_for_userid
    - export_user_data
    - delete_data_for_all_users_in_context
    - delete_data_for_user
    - get_users_in_context
    - delete_data_for_users

in `classes/privacy/provider.php`.

### Example

Here is an example from one of the existing Moodle plugins (Blackboard Collaborate): https://github.com/open-lms-open-source/moodle-mod_collaborate/blob/master/classes/privacy/provider.php

#### Running the "[Test of privacy API compliance](https://docs.moodle.org/dev/Privacy_API/Utilities)" script

Screen Shot 2022-04-14 at 19 51 08

#### From admin page: `/admin/tool/dataprivacy/pluginregistry.php`

Screen Shot 2022-04-15 at 13 55 31

#### Running the "[Test of privacy API compliance](https://docs.moodle.org/dev/Privacy_API/Utilities)" script

Screen Shot 2022-04-14 at 19 51 25

#### From admin page: `/admin/tool/dataprivacy/pluginregistry.php`

Screen Shot 2022-04-15 at 13 21 02

### CODE REVIEW

**Review by octaviorequena — APPROVED** [review]
Works as expected! 🚀 

Screenshot 2022-04-18 at 10 25 27

File: classes/privacy/provider.php [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    /**
+     * Returns meta data about this system.
+     *
+     * @param  collection $collection The collection to add metadata to.
+     * @return collection $collection The array of metadata.
+     */
+    public static function get_metadata(collection $collection) : collection {
+
+        $collection->add_external_location_link('accredible', [
+            'email' => 'privacy:metadata:accredible:email',
```
Comment (alanheppenstall) [inline classes/privacy/provider.php:49]: We also use the grade right? That would be personal data under their definition, so would need to be included in the implementation.

**Review by alanheppenstall — COMMENTED** [review]
Please can we just consider the grade field and any other data we capture from the Moodle plugin?

Looking through the codebase, grade is the only other item I can see that we use.

File: classes/privacy/provider.php (hunk 1 above)
Comment (kei178) [reply classes/privacy/provider.php:49]: I added quiz grade to the metadata: https://github.com/accredible/moodle-mod_accredible/pull/58/commits/5708a043f6ee638114651f280446df10a3cffae8

Screen Shot 2022-04-20 at 10 10 24

File: lang/en/accredible.php [hunk 2]
```diff
@@ -78,3 +78,8 @@
 $string['overview'] = 'Overview';
 $string['activitygroupdescription'] = 'Credentials groups need to have been created in the <a href="{$a}" target="_blank">Accredible Dashboard</a> before credentials can be issued. If none appear, check your API Key to make sure integration is set up correctly.';
 $string['accrediblegroup'] = 'Accredible Group';
+
+$string['privacy:metadata:accredible'] = 'In order to integrate with Accredible, user data needs to be exchanged with that service.';
+$string['privacy:metadata:accredible:email'] = 'Your email address is sent to Accredible to issue a credential.';
+$string['privacy:metadata:accredible:fullname'] = 'Your full name is sent to Accredible to issue a credential.';
+$string['privacy:metadata:accredible:quizgrade'] = 'Your quiz grade is sent to Accredible to issue a credential.';
```
Comment (alanheppenstall) [inline lang/en/accredible.php:85]: Can we reword this to:
Your quiz grade **may be** sent to Accredible to issue a credential.

Comment (kei178) [reply lang/en/accredible.php]: That's a good point as we're also planning to add a new setting option to disable evidence item creation. I updated the wording: https://github.com/accredible/moodle-mod_accredible/pull/58/commits/b673d36a49b5be0b395ed37f9e9fca83633dd648

Screen Shot 2022-04-20 at 18 46 35

**Review by alanheppenstall — APPROVED** [review]
